### ISSUE TITLE
JMS Reader and JMS Writer don't automatically reconnect upon broker restart

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#2141]
Channels that use a JMS Reader connector or JMS Writer connector must be redeployed (or stopped then started) after the ActiveMQ broker it is connected to restarts.

The problem is this: 
When ActiveMQ is restarted (or in my case, software that uses an embedded broker) any channel that is connected to that broker will throw this error:

2012-08-02 07:37:00,343]  ERROR (com.mirth.connect.connectors.jms.JmsMessageReceiver:409): Initialisation Failure: The Consumer is closed
org.mule.umo.lifecycle.LifecycleException: Initialisation Failure: The Consumer is closed
	at com.mirth.connect.connectors.jms.JmsMessageReceiver.doStop(JmsMessageReceiver.java:126)
	at org.mule.providers.AbstractMessageReceiver.stop(AbstractMessageReceiver.java:407)
	at org.mule.providers.AbstractMessageReceiver.dispose(AbstractMessageReceiver.java:306)
	at org.mule.providers.AbstractConnector.destroyReceiver(AbstractConnector.java:588)
	at org.mule.providers.AbstractConnector.unregisterListener(AbstractConnector.java:548)
	at org.mule.impl.model.AbstractModel.unregisterListeners(AbstractModel.java:280)
	at org.mule.impl.model.AbstractModel.stopComponent(AbstractModel.java:478)
	at org.mule.management.mbeans.ModelService.stopComponent(ModelService.java:55)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:616)
	at com.sun.jmx.mbeanserver.StandardMBeanIntrospector.invokeM2(StandardMBeanIntrospector.java:111)
	at com.sun.jmx.mbeanserver.StandardMBeanIntrospector.invokeM2(StandardMBeanIntrospector.java:45)
	at com.sun.jmx.mbeanserver.MBeanIntrospector.invokeM(MBeanIntrospector.java:226)
	at com.sun.jmx.mbeanserver.PerInterface.invoke(PerInterface.java:138)
	at com.sun.jmx.mbeanserver.MBeanSupport.invoke(MBeanSupport.java:251)
	at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:857)
	at com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:795)
	at javax.management.remote.rmi.RMIConnectionImpl.doOperation(RMIConnectionImpl.java:1450)
	at javax.management.remote.rmi.RMIConnectionImpl.access$200(RMIConnectionImpl.java:90)
	at javax.management.remote.rmi.RMIConnectionImpl$PrivilegedOperation.run(RMIConnectionImpl.java:1285)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.management.remote.rmi.RMIConnectionImpl.doPrivilegedOperation(RMIConnectionImpl.java:1390)
	at javax.management.remote.rmi.RMIConnectionImpl.invoke(RMIConnectionImpl.java:807)
	at sun.reflect.GeneratedMethodAccessor13.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:616)
	at sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:322)
	at sun.rmi.transport.Transport$1.run(Transport.java:177)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.rmi.transport.Transport.serviceCall(Transport.java:173)
	at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:553)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:808)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:667)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603)
	at java.lang.Thread.run(Thread.java:679)Caused by: javax.jms.IllegalStateException: The Consumer is closed
	at org.apache.activemq.ActiveMQMessageConsumer.checkClosed(ActiveMQMessageConsumer.java:811)
	at org.apache.activemq.ActiveMQMessageConsumer.setMessageListener(ActiveMQMessageConsumer.java:415)
	at com.mirth.connect.connectors.jms.JmsMessageReceiver.doStop(JmsMessageReceiver.java:123)
	... 37 more

This error appears in the Mirth Connect log but it does not show an error state directly on the channel nor does it send an alert.

The channel never attempts to reconnect and will only resume processing messages if redeployed (or stopped and started).

How to reproduce this error:

1. Setup ActiveMQ (or create an embedded ActiveMQ broker)

2. Create a channel with a JMS Reader connector and/or JMS Writer connector

3. Start ActiveMQ

4. Deploy channel

5. Stop ActiveMQ

6 (Optional). Queue message on ActiveMQ

7 (Optional). Start ActiveMQ (at this moment, the messages will be consumed).

Recommendations: 
The LLP connectors are a great example of how this should work. The connections should be automatically reestablished ever x seconds. Any unsent messages on a JMS writer should be queued up as in the LLP Sender connector. The JMS Reader should simply retry to connect.  

This issue is similar to, but not exactly the same, as http://www.mirthcorp.com/community/issues/browse/MIRTH-1083

Imported Issue. Original Details:
Jira Issue Key: MIRTH-2197
Reporter: mbs
Created: 2012-08-02T12:54:08.000-0700

**Comment by rbeckman-nextgen** [comment]
Definitely a big problem.  JMS channels are unusable with this problem.

Imported Comment. Original Details:
Author: offbyone
Created: 2012-08-02T14:42:19.000-0700

**Comment by rbeckman-nextgen** [comment]
I was able to get it to work. It was partially a problem with ActiveMQ. ActiveMQSslConnectionFactory.setKeyAndTrustManagers is suppose to set the SSL context for the connection, but it doesn't work. Thread.setSslContext does work though.

At this point, my own code could connect without error, but Mirth Connect still caused a certificate unknown error when it connected.

I was able to use a workaround to get Mirth Connect to work with ActiveMQ, but I had to set the keystore, truststore, and passwords (respectively) in the mcservice.vmoptions file (using -Djava.net.ssl.keyStore...). This does work, but it also means that all the JMS channels now must use the same keystore and truststore. Previously I used a per-channel SSL context setup via the Property and Value table in the JMS connectors. The properties in this table did work until I enabled failover (at this point I had to use the "global" workaround).

It would be nice if the JMS connectors had a "Use SSL Context" checkbox with associated textboxes for the keystore and truststore. But, at the very least, the Property and Value table should work in this manner.

Imported Comment. Original Details:
Author: mbs
Created: 2012-08-23T10:18:44.000-0700

**Comment by rbeckman-nextgen** [comment]
The JMS connectors have been rewritten in version 3.0.0 Beta 2. They now include a feature to automatically attempt to reconnect when the connection to the JMS broker goes down.

Imported Comment. Original Details:
Author: brentm
Created: 2013-01-30T11:40:52.000-0800

**Comment by rbeckman-nextgen** [comment]
Confirmed that if the JMS Sender can't connect to the broker, it simply queues the message. One thing to note is that the reconnection seems to happen asynchronously, meaning that if queuing is disabled on the destination and the broker gets restarted, any messages that get sent in the meantime will be errored out, instead of waiting for the reconnection to finish. I created a separate issue for that though: MIRTH-2447

On the JMS Reader side, since the reconnect happens automatically there are really no problems. Once the connection is eventually established, any messages will flow through as normal.

Imported Comment. Original Details:
Author: narupley
Created: 2013-04-01T12:51:56.000-0700
